### ISSUE TITLE
advanced seed for community in progress - request for assistance

### ISSUE THREAD

**Issue by NathanWalker** [NativeScript/nativescript-angular#60]
I am working on a relatively advanced angular2-seed which will feature {N} integration:
https://github.com/NathanWalker/angular2-seed-advanced

I'm running into an issue. Curious if someone could be so kind to take a look by following the instructions here to get setup and trying to help:
https://github.com/NathanWalker/angular2-seed-advanced/blob/development/NATIVESCRIPT_SUPPORT.md

**Comment by hdeshev** [comment]
`JS ERROR ReferenceError: Can't find variable: __metadata`

You get this one if you don't have `reflect-metadata` imported, or your compiler isn't emitting helpers or decorator metadata.

Out of curiosity, why do users need to edit tns-core-modules.d.ts? What is the original issue, and can we do anything about it?

**Comment by NathanWalker** [comment]
`tns-core-modules.d.ts` includes references to:
```
/// <reference path="declarations.d.ts" />
/// <reference path="es-collections.d.ts" />
/// <reference path="es6-promise.d.ts" />
/// <reference path="es6.d.ts" />
/// <reference path="weakmap.d.ts" />
```

Those cause problems when paired with other build setups that already include those typings.
We should probably get NativeScript typings on the [typings registry here](https://github.com/typings/registry) since `tsd` is deprecated.
Instead of including those global `typings` in `tns-core-modules.d.ts`, it could be instructed for end user's to setup their `typings.json` to contain the `typings` that are needed. 
This way `tns-core-modules.d.ts` would **only** cover `typings` for NativeScript libraries.

**Comment by NathanWalker** [comment]
I am importing `reflect-metadata` here:
https://github.com/NathanWalker/angular2-seed-advanced/blob/development/src/app/app.ts
Maybe it needs to be imported in other files as well?
Maybe something incorrect here:
https://github.com/NathanWalker/angular2-seed-advanced/blob/development/src/tsconfig.json

Thank you for your help @hdeshev

**Comment by NathanWalker** [comment]
I was able to get past it by removing both of these options in `tsconfig.json`:
```
"noEmitHelpers": true,
        "noEmitOnError": true,
```

:+1:  It's close to working now... just a few things left to work with `Http` and `ng2-translate`

**Comment by NathanWalker** [comment]
Ok @hdeshev I've gotten much further with your help (thank you!) and I updated the {N} instructions here:
https://github.com/NathanWalker/angular2-seed-advanced/blob/development/NATIVESCRIPT_SUPPORT.md

But I am now running into this:
```
CONSOLE ERROR file:///app/tns_modules/angular2/src/platform/server/parse5_adapter.js:53:75: ORIGINAL EXCEPTION: TypeError: this.http.get(this.sfLoaderParams.prefix + "/" + lang + this.sfLoaderParams.suffix)
	            .map is not a function. (In 'this.http.get(this.sfLoaderParams.prefix + "/" + lang + this.sfLoaderParams.suffix)
	            .map(function (res) { return res.json(); })', 'this.http.get(this.sfLoaderParams.prefix + "/" + lang + this.sfLoaderParams.suffix)
	            .map' is undefined)
```

This is coming from the internals of `ng2-translate` (https://github.com/ocombe/ng2-translate/blob/master/src/translate.service.ts#L40-L41).

I have of course imported the `.map` operator here:
https://github.com/NathanWalker/angular2-seed-advanced/blob/development/src/app/app.ts#L7

I have also seen your `Http` sample here:
https://github.com/NativeScript/sample-ng-http

But I don't see anything I'm missing. Any idea what this last error is and why it's not importing the `.map` operator correctly?

**Comment by NathanWalker** [comment]
Apparently this same question was asked on stackoverflow a couple weeks ago with no answer:
http://stackoverflow.com/questions/34899835/using-map-in-nativescript-with-angular2

I am experiencing the same trouble.

**Comment by ligaz** [comment]
Hey @NathanWalker,

The issue looks like this one https://github.com/NativeScript/nativescript-angular/issues/38 which has already been fixed in the iOS runtime. You can try downloading the latest iOS runtime bits from [here](http://nativescript.rocks/) and use them. As those builds are not verified I'm not sure what will be their stability.

**Comment by NathanWalker** [comment]
@ligaz Thank you! I will try this! ..crossing fingers..

**Comment by NathanWalker** [comment]
Hmm unfortunately I now get this error setting up all the latest nightly's as that nativescript.rocks site mentions:
```
ld: warning: directory not found for option '-F/Users/nathan/Documents/github/NathanWalker/angular2-seed-advanced/src/platforms/ios/../../lib/iOS'
ld: can't open -sectcreate file: /Users/nathan/Documents/github/NathanWalker/angular2-seed-advanced/src/platforms/ios/build/emulator/metadata-i386.bin
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by hdeshev** [comment]
The typings problem is related to NativeScript/NativeScript#1457. I started working on it today, and it will probably land in master tomorrow. The idea is for Angular apps to reference a different dts file for the nativescript components. I'm not sure `tns-core-modules.es6.d.ts` will suffice yet -- I may have to come up with another :-/

Anyway, solving this problem would get rid of the ugly build hook that regex-replaces stuff in the angular d.ts files to delete certain references that cause compile-time type conflicts.

**Comment by NathanWalker** [comment]
Ok that makes sense. Thank you for this insight and let me know if I can help in anyway to move that forward. Give me a task, seriously. I'm quite anxious to get this working :)

**Comment by NathanaelA** [comment]
@ligaz  - The builds on NativeScript.rocks are straight from source every afternoon my time.   (The majority of the NS core team is deeply sleeping midday my time; so it is the safest time to make sure I capture everyone's work on master for the day).    

The actual status:
1. Android runtimes have my liveedit & encryption patches automatically applied.
2. iOS runtimes, no patches yet -- but eventually will have the same patches as above.
3. TNS Core Master - totally unchanged, just TS -> JS'd. 

These are all build from Master, including the required sub-projects (widgets, metadata, v8).  
v8 is only built when a new v8 is setup in master; as their is no point rebuilding the same code everyday --- so right now my masters are actually running on the new 4.7.80.  :grinning: 

I do occasionally have to kick the android build as occasionally the NS team breaks my patches by adding new code around one of the diff locations and diff's won't apply cleanly, but otherwise it is a totally automated process; so if it doesn't work; it is because the master is broken.  :grinning:

**Comment by tjvantoll** [comment]
I ran into this today as well so I thought I’d just add two things:

- 1) HTTP calls work for me on Android but not iOS, which tends to suggest that the problem is indeed the iOS runtime bug that @ligaz suggests.
- 2) Oddly, I had to do a `JSON.parse()` step to actually get my data in JSON format. See https://github.com/NativeScript/sample-Groceries/blob/bcb03dc54e8c4575b17fb99a6a22e8857b184a5c/app/shared/user/user.service.ts#L26. Anybody know what might be up there?

The good news: other than that one line of code, I’m using the _exact_ same service to drive my web app and my native app, and that’s pretty freakin’ awesome 😀
